design/src/rfcs/rfc0016_body_callback_apis_v2.md

```diff
@@ -0,0 +1,389 @@
+        } else if checksum_algorithm.eq_ignore_ascii_case(CRC_32_C_NAME) {
+            Self(Inner::Crc32c(Crc32cCallback::default()))
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_1_NAME) {
+            Self(Inner::Sha1(Sha1Callback::default()))
+        } else if checksum_algorithm.eq_ignore_ascii_case(SHA_256_NAME) {
+            Self(Inner::Sha256(Sha256Callback::default()))
+        } else {
+            panic!("unsupported checksum algorithm '{}'", checksum_algorithm)
+        }
+    }
+
+    pub fn update(&mut self, bytes: &[u8]) -> Result<(), BoxError> {
+        match &mut self.0 {
+            Inner::Crc32(ref mut callback) => callback.update(bytes)?,
+            Inner::Crc32c(ref mut callback) => callback.update(bytes)?,
+            Inner::Sha1(ref mut callback) => callback.update(bytes)?,
+            Inner::Sha256(ref mut callback) => callback.update(bytes)?,
+        };
+
+        Ok(())
+    }
+
+    pub fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> {
+        match &self.0 {
+            Inner::Sha256(callback) => callback.trailers(),
+            Inner::Crc32c(callback) => callback.trailers(),
+            Inner::Crc32(callback) => callback.trailers(),
+            Inner::Sha1(callback) => callback.trailers(),
+        }
+    }
+
+    // TODO I think I still need to account for the size of the trailer names.
+    //      Should I be doing that here?
+    pub fn size_hint(&self) -> SizeHint {
+        let checksum_size_in_bytes = match &self.0 {
+            Inner::Crc32(_) => Crc32Callback::size(),
+            Inner::Crc32c(_) => Crc32cCallback::size(),
+            Inner::Sha1(_) => Sha1Callback::size(),
+            Inner::Sha256(_) => Sha256Callback::size(),
+        };
+        let base64_encoded_checksum_size_in_bytes =
+            base64::encoded_length(checksum_size_in_bytes as usize);
+
+        SizeHint::with_exact(base64_encoded_checksum_size_in_bytes as u64)
+    }
+}
+```
+
+```rust
+// In aws-smithy-checksums/src/body.rs
+use crate::ChecksumCallback;
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::header::append_merge_header_maps;
+
+use bytes::{Buf, Bytes};
+use http::{HeaderMap, HeaderValue};
+use http_body::{Body, SizeHint};
+use pin_project::pin_project;
+
+use std::error::Error as StdError;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+#[pin_project]
+pub struct ChecksumBody<InnerBody> {
+    #[pin]
+    inner: InnerBody,
+    #[pin]
+    checksum_callback: ChecksumCallback,
+}
+
+impl ChecksumBody<SdkBody> {
+    pub fn new(checksum_algorithm: &str, body: SdkBody) -> Self {
+        Self {
+            checksum_callback: ChecksumCallback::new(checksum_algorithm),
+            inner: body,
+        }
+    }
+
+    // TODO a more intelligent person could prob impl this with generics
+    fn poll_inner(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Bytes, Error>>> {
```

Comment:
What would be gained by using generics for this function?

Reply:
Then it would support other body types by default. I also considered parameterizing this on the `http_body::Body` trait so that it'd be more generic. If y'all aren't concerned then I won't be either.

After thinking about it, I think my understanding/question is:
"We make wrapper structs generic so that we can support other wrapped types in the future. Is it better to implement the wrappers with concrete types as necessary or should we always seek to parameterize on traits that wrapped types must implement?"

Reply:
My thinking on this is that `SdkBody` isn't generic, so the thing that wraps `SdkBody` doesn't need to be generic. Given how infectious generics are in Rust, I'm inclined to index on using less of them rather than making things overly generic without a concrete use-case.

Reply:
I think that makes a lot of sense. 👍